### PR TITLE
Add `extract_location` preprocessor to extract town, city, mountains etc - anything specifiable by a location

### LINKED ISSUE #1334: Add support for extracting a specific point by location name
It would be nice to have the option to specify the point we want to extract with just the location name and not the coordinates. It will save most users one search and will make recipes more readable.

So for example, this preprocessor from the `recipe_python.yaml`
```yaml
  annual_mean_amsterdam:
    extract_point:
      latitude: 52.379189
      longitude: 4.899431
      scheme: linear
    annual_statistics:
      operator: mean
    multi_model_statistics:
      statistics:
        - mean
      span: overlap
```
will be just

```yaml
  annual_mean_amsterdam:
    extract_location:
      location: Amsterdam
      scheme: linear
    annual_statistics:
      operator: mean
    multi_model_statistics:
      statistics:
        - mean
      span: overlap
```

There is already a PR open in #1251 with implements this using openstreetmap nominatim through geopy and delegates the extraction itself to the `extract_point` preprocessor

### PR DESCRIPTION
**Pull request by jvegreg**
This PR adds a new preprocessor that is able to extract a point by city name. It justs retrieve the lat lon coordinates using Geopy and calls the `extract_point` preprocessor.

@valeriupredoi says: I am taking over this PR so we can get it merged and into 2.5.0 - `tis a nice feature :mountain: 

It is now ready for review; I have done the following to it:
- added a bit more docstrings (Raises etc)
- I moved `geopy` from PyPi to conda-forge (for us I mean) https://anaconda.org/conda-forge/geopy - tests run fine, see [Github Action](https://github.com/ESMValGroup/ESMValCore/runs/4871197861?check_suite_focus=true)
- I added documentation

Closes #1334

Link to documentation: https://esmvaltool--1251.org.readthedocs.build/projects/ESMValCore/en/1251/recipe/preprocessor.html#extract-location

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

### CODE REVIEW

**Comment by valeriupredoi** [comment]
I reckon it's probably better if you name it `extract_city` or `extract_urban_area` man - town is rather very specific and small in size too :city_sunset:

**Comment by jvegasbsc** [comment]
> I reckon it's probably better if you name it extract_city or extract_urban_area man - town is rather very specific and small in size too 🌆

I just found that it also works for mountains and other geographical stuff, so I will use location instead

**Comment by zklaus** [comment]
I think the dependency needs to be added to a few more places. Have a look at https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#dependencies. I am surprised to find it in `doc/conf.py`, though, but maybe that's something to discuss in the proper review.

Could you please also link the issue with the preceding discussion of the feature?

**Comment by jvegasbsc** [comment]
> Could you please also link the issue with the preceding discussion of the feature?

There was not such issue, but it is now created #1334 

> I am surprised to find it in doc/conf.py, though, but maybe that's something to discuss in the proper review.

It is just because I misplaced that dependency there instead of the requirements.txt

**Comment by valeriupredoi** [comment]
`extract_24h_Le_Mans_racetrack` and I'll buy you a month's supply of :beer:

File: esmvalcore/preprocessor/_regrid.py
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_regrid.py:336]: add a `Raises` section so the user knows there's an exc is raised if they enter mythical places :+1:

File: esmvalcore/preprocessor/_regrid.py
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_regrid.py:340]:
```suggestion
        raise ValueError(f'Location {location} can not be found.')
```

File: esmvalcore/preprocessor/_regrid.py
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_regrid.py:343]: are you letting the code crash if no `scheme` is provided?

File: tests/integration/preprocessor/_regrid/test_extract_location.py
Comment (valeriupredoi) [inline tests/integration/preprocessor/_regrid/test_extract_location.py:110]: hahahaha :rofl:

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
cheers Javi! Couple minor comments from me :+1:

**Comment by valeriupredoi** [comment]
I'll try migrate `geopy` from PyPi to conda-forge (for us I mean) https://anaconda.org/conda-forge/geopy

**Review by valeriupredoi — APPROVED** [review]

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:347]:
```suggestion
    logger.debug("Extracting data for %s (%s ºN, %s ºW)", location,
```

Usually we don't show messages of individual preprocessors in the info log.

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:341]: This probably also raises an `ValueError` if `scheme` is invalid.

File: tests/integration/preprocessor/_regrid/test_extract_location.py
Comment (schlunma) [inline tests/integration/preprocessor/_regrid/test_extract_location.py:2]:
```suggestion
"""Integration tests for :func:`esmvalcore.preprocessor.extract_location."""
```

File: tests/integration/preprocessor/_regrid/test_extract_location.py
Comment (schlunma) [inline tests/integration/preprocessor/_regrid/test_extract_location.py:110]: :ring:

**Review by schlunma — COMMENTED** [review]
Code looks good, just a couple of small comments on documentation.

Will test now with real data!

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:346]: You need to assign another variable name for `location`, if the location is not find I get the following error:

```
ValueError: Requested location None can not be found.
```

And maybe check for `if ... is not None` instead of `if not ...`?

**Comment by schlunma** [comment]
I successfully tested this with many examples (cities, states, countries, lakes, mountains, etc.). This works exceptionally well! :tada: 

I also got this nice message for `location: mount everest`:

```
Extracting data for 珠穆朗玛峰/सगरमाथा, Khumjung, Khumbupasanglahmu, सोलुखुम्बु, Province #1, नेपाल (27.9882393 ºN, 86.925017 ºW)
```


Apart from that I realized that the error message does not make sense if the location is not found, if this is fixed I will approve!

**Comment by valeriupredoi** [comment]
many thanks for the review and tests @schlunma - I'll plug them suggestions in asap today :beer:

File: doc/recipe/preprocessor.rst
Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:1488]: could you maybe give me the locations you tested with so I can list them as examples please? No lists asa far as I can see, prob best to keep it this way, otherwise it'd be called `extract_locations`?

Comment (schlunma) [reply doc/recipe/preprocessor.rst:1488]: I tested `'ammersee'`, `'münchen'`, `'frankfurt'`, `'frankfurt, oder'`, `'usa'`, `'new york, usa'`, `'watzmann'`, `'mount everest'`. Just pick something you like :smile:

File: esmvalcore/preprocessor/_regrid.py
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:341]: yeah but that's taken care of from `extract_point` - that one raises its own ValueError, I checked

File: tests/integration/preprocessor/_regrid/test_extract_location.py
Comment (valeriupredoi) [reply tests/integration/preprocessor/_regrid/test_extract_location.py:109]: :hobbit:

File: esmvalcore/preprocessor/_regrid.py
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:346]: ahem, will do, a check if `None` is prob enough here

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:341]: I know, but might make sense to add it here since people might not immediately connect the two preprocessors. However, up to you.

**Comment by valeriupredoi** [comment]
that location written in Nepali is awesome :grin: Or is it Nepalese? :nepal:

File: doc/recipe/preprocessor.rst
Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:1488]: Nice! US is a bit bigger than a point, I wonder what actual point it used, Daytona :racing_car: :checkered_flag:

Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:1488]: OK I added examples :wink:

File: esmvalcore/preprocessor/_regrid.py
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py]: done!

File: esmvalcore/preprocessor/_regrid.py
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:346]: sure I added a ValuError here if not supplied, before extract point catches it, but if supplied and not either linear or nearest then extract point catches it

**Comment by valeriupredoi** [comment]
@schlunma I made the changes that should bring this up to full running order, cheers for suggestions, have another look pleae and :+1: or :-1:

File: doc/recipe/preprocessor.rst
Comment (schlunma) [inline doc/recipe/preprocessor.rst:1493]:
```suggestion
    'romania', 'new york, usa'. Raises ValueError if none supplied.
```

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:346]: I was just referring to the documentation, I didn't want you to add a check in the code :grimacing:

Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:346]:
```suggestion
    ValueError:
        If given location cannot be found by the geolocator.
```

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:358]: Ah, sorry, I wasn't specific enough:

If the location cannot be found, `location` is set to `None` in l.357. This leads to the error message `ValueError: Requested location None can not be found.`, regardless of the location specified. I would like to have a message like `ValueError: Requested location Gondor can not be found.`

**Comment by schlunma** [comment]
> btw the check if it's None before that is also useful, the recipe parser populates the mandatory preprocessor parameters with None values if not user-specified

No, I'm pretty sure it doesn't. If there is no default for the variable in the function definition, it the recipe parser raises an error if a parameter is missing.

**Comment by valeriupredoi** [comment]
you're right! l.236 in `preprocessor/__init__.py` phew :grin: Anyways, it could be `location: Null` in recipe no?

**Comment by schlunma** [comment]
True, but that would probably raise an error in `geolocator.geocode(location)` :grin:

**Comment by valeriupredoi** [comment]
> True, but that would probably raise an error in `geolocator.geocode(location)` grin

just catching it before, man, I like checks :grin:

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:362]:
```suggestion
    geolocation = geolocator.geocode(location)
    if geolocation is None:
        raise ValueError(f'Requested location {location} can not be found.')
    logger.debug("Extracting data for %s (%s ºN, %s ºW)", geolocation,
                 geolocation.latitude, geolocation.longitude)

    return extract_point(cube, geolocation.latitude, geolocation.longitude, scheme)
```

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:362]: why do you want to rename the variable? if `location` returned by geolocator is None, then it raises, if not it'll go and do the extraction

Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:362]: Because of the error message: Just try a wrong location with the current version of this branch, e.g. `location: Gondor`. This will give you the error:

```
ValueError: Requested location None can not be found.
```

But I want

```
ValueError: Requested location Gondor can not be found.
```

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:362]: ah gotcha! The raise error needs to have the actual input location :man_facepalming: Sorry, am slow today

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:359]:
```suggestion
    logger.debug("Extracting data for %s (%s °N, %s °E)", geolocation,
```

Sorry, my latest edits weren't included, this needs to be °E, not °W

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py]: no idea why E now W but I'll trust you, I don't like East though :laughing:

Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py]: Well for a place in Germany I got `50N 8W`, and Germany is clearly east of London :grimacing:

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py]: as it stands, it's way more West in terms of mentality :rofl:

**Comment by schlunma** [comment]
Everything works fine now, thanks :tada: 

Just one more thing (I promise, this will be the last): This feature requires a internet connection; if there is non it will raise an

```
geopy.exc.GeocoderUnavailable: HTTPSConnectionPool(host='nominatim.openstreetmap.org', port=443): Max retries exceeded with url: /search?q=denver%2C+usa&format=json&limit=1 (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x2acd1528e530>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

Do we need to mention this in the doc?

**Comment by valeriupredoi** [comment]
I'll be a sea urchin! How'd you figure that out? had no idea - yeh, let me write it up in the dox, very good find :+1:

**Comment by schlunma** [comment]
Most of the computing nodes on mistral are not connected to the open internet, so this is actually not an unrealistic scenario.

File: doc/recipe/preprocessor.rst
Comment (schlunma) [inline doc/recipe/preprocessor.rst:1491]:
```suggestion
.. note::
   Note that this function's geolocator application needs a
   working internet connection.
```

File: esmvalcore/preprocessor/_regrid.py
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:322]:
```suggestion
    Note
    ----
    The geolocator needs a working internet connection.
```

File: doc/recipe/preprocessor.rst
Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:1491]: yeh OK lots other functions have in-line notes

**Comment by valeriupredoi** [comment]
> > Most of the computing nodes on mistral are not connected to the open internet, so this is actually not an unrealistic scenario.
> 
> Sorry to step up, but that is part of the reason I was showing the actual location in an info message. You can run a very small set in your laptop to get the coordinates and then switch to extract point in the big machine.
> 
> It was also an info message for the case of homonyms. There are lot's of places that share the name and I wanted to show the actual one that was found to avoid strange mistakes

hey look who's back :grin: :es: :tada: 

I know, there is Rome, New York (state), and actual Rome :grin: Do you want to add more stuffs before we merge, Javi?

**Comment by schlunma** [comment]
Agreed, that makes sense. But if you have lots of datasets you will get that message very often, and i thought this is something we don't want. Maybe there is a smarter way of doing this? Run the search already in `_recipe.py` and pass the `geolocator` object to the preprocessor (after making sure that the function accepts this, of course)?

**Comment by valeriupredoi** [comment]
naah, that's too complicated an approach, I say let the user figure it out :smiling_imp:

**Comment by valeriupredoi** [comment]
> I don't think it's that complicated, just add an additional function here:
> 
> https://github.com/ESMValGroup/ESMValCore/blob/e9daad3b648d4872bfd67320b68d578faa1faee5/esmvalcore/_recipe.py#L903
> 
> I can give it a try if you like smile

no I don't think I'd like that, `_recipe.py` is a glorified parser and manager, we should avoid running any analysis inside it, calls to functions that run stuff outside it yes, but not any actual running inside it, man

**Comment by schlunma** [comment]
I wouldn't run any analysis in there, just retrieve the geolocation object and pass it as option to the preprocessor. We actually do stuff like that very often, e.g., retrieve fx files needed as input for the preprocessors or convert `levels` given as CMOR levels to something that the `extract_levels` preprocessor can handle.

If we don't do this, then I would actually promote the `debug` message to an `info` message again, @jvegreg has a point here.

**Comment by valeriupredoi** [comment]
ah I thought you wanted to extract the location there - I think promoting the message to `info` might be better - well, the best way would be if the geolocator could return some sort of warning a la "found multiple locations with same name" and we'd raise that - got no idea about that thing tho

**Comment by schlunma** [comment]
All right, fine for me!

File: doc/recipe/preprocessor.rst
Comment (schlunma) [inline doc/recipe/preprocessor.rst:1476]:
```suggestion
    ``'nearest'``. There is no default.
    
See also :func:`esmvalcore.preprocessor.extract_point`.
```

File: doc/recipe/preprocessor.rst
Comment (schlunma) [inline doc/recipe/preprocessor.rst:1499]:
```suggestion
    There is no default, raises ValuError if none supplied.

See also :func:`esmvalcore.preprocessor.extract_location`.
```

Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst]: fixed off-site since I also had to fix ValuError -> ValueError (for the 68th time :rofl: )

**Comment by valeriupredoi** [comment]
cheers Manu! This has been a most polished PR, true valu. Value! I mean :rofl:

**Review by schlunma — APPROVED** [review]
Thanks V, great stuff! :beers:

**Comment by valeriupredoi** [comment]
awesome, cheers, bud! :beers:

**Comment by zklaus** [comment]
Very cool. I would like to nominate this feature as a highlight for 2.5.0.

Release managers @schlunma, @axel-lauer, and others: how do you want to manage highlights? Last time we tried an issue, but that didn't attract too much activity and was not so easy to keep informative. How about we try a discussion this time. That way, every nomination can be one answer, people can vote on the answers and the order will reflect the votes. Then we can take the top n features up for the highlight list. What do you think?

**Comment by schlunma** [comment]
Great idea! I opened ESMValGroup/ESMValTool#2495 for that!

**Comment by zklaus** [comment]
Cheers @schlunma. I made some slight edits to that to clarify that this is for ESMValCore only and to say that there should be only exactly one, already merged PR per answer. If you disagree with that, feel free to roll-back the edit.

**Comment by valeriupredoi** [comment]
grilling-out

**Pull request merged**
